### PR TITLE
Sw 57/adjustable grid

### PR DESCRIPTION
**Pull request by AdityaKompella6**
https://cornell-nexus.atlassian.net/browse/SW-57?atlOrigin=eyJpIjoiMTg3ZDBmZTAzOTMwNGUxN2E0MjY1MmY5ODU5NjlhZWMiLCJwIjoiaiJ9

### CODE REVIEW

**Comment by sazhang02** [comment]
"I created a script to test the adjustable grid called test_activations_shape.py

If you run this file, it will bring up a prompt where you can define a rectangular grid that is num_rows*num_cols. You can then activate a circle inside it(defined by center and radius) or a triangle(defined by 3 (x,y) pairs) to see the lawnmower traversal. There are comments that suggest good points/radius for triangle/circle activation. Green nodes are activated nodes, the red nodes are border nodes and the purple lines are the waypoints the lawnmower algorithm traverses. the algorithm starts from the top-right most activated point and goes all the way to the left and then circles back to the right in the next row down to cover all the activated nodes.

To get the full functionality, have to call activate_rectangle/circle/triangle with chosen parameters and then run get_all_lawnmower_waypoints_adjustable to get the waypoints." - Summary copied from Jira ticket for reference

File: engine/grid.py [hunk 1]
```diff
@@ -185,27 +185,128 @@ def is_on_border(self, row, col, row_limit, col_limit):
...
+        """
+        rows = self.nodes.shape[0]
+        cols = self.nodes.shape[0]
+        for x in range(rows):
+            for y in range(cols):
+                if ((x-circle_center_row)**2 + (y-circle_center_col)**2 - circle_radius**2) < 0:
+                    self.nodes[x,y].is_active = True
+
+    def area(x1, y1, x2, y2, x3, y3):
+
```
Comment (sazhang02) [inline engine/grid.py:231]:
```suggestion
```
delete space

File: engine/test_activation_shapes.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+    if activation_type== "circle":
+        circle_center_row = int(input("Circle center row: "))## Good value 15
+        circle_center_col = int(input("Circle center col: "))## Good value 15
+        circle_radius = int(input("Circle radius: "))## Good value 12
+        nodes = activate_circle(nodes, circle_center_row, circle_center_col, circle_radius)  
+        border_nodes,_,leftmost_node_pos, nodes = find_border_nodes(nodes)
+        way_points = get_all_lawnmower_waypoints_adjustable(nodes,leftmost_node_pos,border_nodes)
+
+    if activation_type== "triangle":
```
Comment (sazhang02) [inline engine/test_activation_shapes.py:194]: lol made a straight line here w/ the triangle type, just checking if this behavior is intended to be ok?

inputs:
Num rows: 30
Num cols: 30
Activation type: triangle
x1: 1
y1: 2
x2: 15
y2: 16
x3: 29
y3: 30
Screen Shot 2022-09-14 at 1 08 08 AM

File: engine/test_activation_shapes.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    plt.title("Activated Nodes")
+    plt.xlim(-1, rows+1)
+    plt.ylim(-1, cols+1)
+    plt.grid()
+    nodes = np.empty([rows, cols], dtype=object)
+    for i in range(rows):
+        for j in range(cols):
+            nodes[i,j] = Node(False,i,j)
+            
+    if activation_type == "rectangle":
```
Comment (sazhang02) [inline engine/test_activation_shapes.py:177]: Should we be able to make straight lines here? crashes if i do
Inputs:
Num rows: 30
Num cols: 30
Activation type: rectangle
Rectangle row start: 5
Rectangle row end: 5
Rectangle col start: 9
Rectangle col end: 9

Comment (sazhang02) [reply engine/test_activation_shapes.py:177]: Also, I tried to make a small rectangle and it highlighted a single border node instead. Is this supposed to happen? Was expecting 4 red dots (but it seems end row is exclusive?)

Inputs:
Num rows: 30
Num cols: 30
Activation type: rectangle
Rectangle row start: 5
Rectangle row end: 6
Rectangle col start: 7
Rectangle col end: 8

Screen Shot 2022-09-14 at 1 11 18 AM

File: engine/test_activation_shapes.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
+import matplotlib.pyplot as plt
```
Comment (sazhang02) [inline engine/test_activation_shapes.py:1]: Lets move this file into the tests.functionality_tests folder

**Review by sazhang02 — CHANGES_REQUESTED** [review]
AMAZING JOB!!!!! Code is very clean, comments and specifications in the code are great + well worded imo, and test script is super cool. I think the script is a wonderful example of a functionality test as it tests this feature end-to-end through visual testing. Added comments w/ some change suggestions.

File: engine/test_activation_shapes.py (hunk 2 above)
Comment (AdityaKompella6) [reply engine/test_activation_shapes.py:194]: Yeah  think this is intended

File: engine/grid.py [hunk 5]
```diff
@@ -185,27 +185,128 @@ def is_on_border(self, row, col, row_limit, col_limit):
...
+        cols = self.nodes.shape[0]
+        for x in range(rows):
+            for y in range(cols):
+                if ((x-circle_center_row)**2 + (y-circle_center_col)**2 - circle_radius**2) < 0:
+                    self.nodes[x,y].is_active = True
+
+    def area(x1, y1, x2, y2, x3, y3):
+    
+        return abs((x1 * (y2 - y3) + x2 * (y3 - y1)
+                    + x3 * (y1 - y2)) / 2.0)
```
[the remark this answers is not in the document]
Comment (AdityaKompella6) [reply engine/grid.py:233]: Done

File: engine/grid.py [hunk 6]
```diff
@@ -214,73 +315,51 @@ def find_border_nodes(self):
...
-        if not neighbor_node.is_active_node():
+    
+    ##Return bottom most node that is activated in the right column
+    def return_right_position(self, pos):
+        candidate_nodes = [node for node in self.border_nodes if node.y == pos[1]+1]
+        if (candidate_nodes == []):
             return None
         else:
-           return neighbor_node
+            node = min(candidate_nodes,key=lambda node: node.x)
```
[the remark this answers is not in the document]
Comment (AdityaKompella6) [reply engine/grid.py:325]: Cool

File: engine/grid.py [hunk 7]
```diff
@@ -214,73 +315,51 @@ def find_border_nodes(self):
...
-        """
-        if row < 0 or row >= row_max or col < 0 or col >= col_max:
-            return None
-
-        node_list = self.nodes
-        neighbor_node = node_list[row][col]
-        if not neighbor_node.is_active_node():
+    
+    ##Return bottom most node that is activated in the right column
+    def return_right_position(self, pos):
```
[the remark this answers is not in the document]
Comment (AdityaKompella6) [reply engine/grid.py:320]: Done

File: engine/test_activation_shapes.py [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+                        phase = WaypointPhase.TERMINATE
+        return waypoints
+
+           
+
+
+
+def main():
+    rows = int(input("Num rows: "))##Good value 30
+    cols = int(input("Num cols: "))## Good value 30
```
[the remark this answers is not in the document]
Comment (AdityaKompella6) [reply engine/test_activation_shapes.py:165]: Um kinda dont handle edge cases

File: engine/test_activation_shapes.py [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+    activation_type = input("Activation type: ")
```
[the remark this answers is not in the document]
Comment (AdityaKompella6) [reply engine/test_activation_shapes.py:166]: Uhhhh I gave some suggestions so I think its fine.

File: engine/test_activation_shapes.py (hunk 3 above)
Comment (AdityaKompella6) [reply engine/test_activation_shapes.py:177]: I think It does but my plotting doesn't

File: engine/test_activation_shapes.py (hunk 4 above)
Comment (AdityaKompella6) [reply engine/test_activation_shapes.py:1]: Done

File: engine/test_activation_shapes.py [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+
+class Node:
+    def __init__(self,activated,x,y):
+        self.activated = activated
+        self.on_border = False
+        self.x = x
+        self.y = y
+
+
+def activate_rectangle(nodes, row, col, row_limit, col_limit):
```
[the remark this answers is not in the document]
Comment (sazhang02) [reply engine/test_activation_shapes.py:13]: lets do this in a diff pr :)

**Review by sazhang02 — APPROVED** [review]

**Pull request merged**